# Python Data Types

**變數型態**:

整數、小數、複數、文字、序列、字典、集合...等，不同的變數型態可使用運算的方式不同。

例：
字串不可用加法： 我 + 你 = ?

Python常用變數型態有14種：

int, float, complex, str, list, tuple, dict, set, range, frozenset, bool, bytes, bytearray and memoryview.

每一種型態皆有不同的運算方式，本篇筆記中將紀錄int, float, complex, str, list, tuple, dict, set, range, bool如何使用。

In [1]:
# 取得變數型態
x = 5
print('Type of x: {}'.format(type(x)))
y = 2+1j
print('Type of y: {}'.format(type(y)))

Type of x: <class 'int'>
Type of y: <class 'complex'>


**備註**：有關print()中的文字，在字串筆記會表示解釋。

# Python Number

**數值型態**:

可用於進行數值運算的變數型態。

- int integer, 整數
 - 可為正數、負數
- float , 浮點數
 - 含小數之數值，皆為浮點數
- complex , 複數
 - 含虛數

In [2]:
x1 = 1
print(type(x1))
x2 = -1
print(type(x2))

y1 = 2.8
print(type(y1))
y2 = -2.7
print(type(y2))

z1 = 1j
print(type(z1))
z2 = 5+7j
print(type(z2))

<class 'int'>
<class 'int'>
<class 'float'>
<class 'float'>
<class 'complex'>
<class 'complex'>


In [3]:
xxx1 = 2E5
print(xxx1)
print(type(xxx1))

200000.0
<class 'float'>


In [4]:
xxx2 = 5e100
print(xxx2)
print(type(xxx2))

5e+100
<class 'float'>


In [5]:
yyy1 = -23.8E5
print(yyy1)
print(type(yyy1))

-2380000.0
<class 'float'>


In [6]:
yyy2 = -87.7e100
print(yyy2)
print(type(yyy2))

-8.77e+101
<class 'float'>


## 浮點數運算誤差

在電腦領域中，所有的「值」皆會使用「二進制」 (0, 1) 表示與運算。

文字：以0, 1組成的代碼表示

數值：以近似值表示，例如：$0.1 = \frac{1}{16}+\frac{1}{32}+\frac{1}{256}+\frac{1}{512}+...$

$0.1 = (\frac{1}{2})^{-4}+(\frac{1}{2})^{-5}+(\frac{1}{2})^{-8}+(\frac{1}{2})^{-9}$

In [7]:
2.4 - 1.3

1.0999999999999999

In [8]:
1 / 10

0.1

### 解決辦法

使用round()函數，進行四捨五入運算。

備註：函數為將一系列程式運算包裝在一起，透過一個名稱取代運算過程。將在「函數」章節筆記。

In [9]:
round(2.4 - 1.3, 1)

1.1

In [10]:
0.1 + 0.1 + 0.1

0.30000000000000004

In [11]:
round(0.1 + 0.1 + 0.1, 1)

0.3

# Python String

**字串型態**:

用於記錄文字內容的型態。

Python中的字串，給定值時需透過一組單引號「'」或雙引號「"」框起來，以表示引號內為「值」而非「變數」。

Python中也支援透過一組連續三單/雙引號「'」或「"」，將一組多行字的字串變成一字串變數。

In [12]:
x = "Hello" # x為內容為 Hello 的變數
print(x)

Hello


In [13]:
# 單個引號組不能分行表示字串內容
x = "Hello 
World
!
"
print(x)

SyntaxError: EOL while scanning string literal (<ipython-input-13-60a7e23f21dc>, line 2)

In [14]:
x = """
Hello
World
!
"""
print(x)


Hello
World
!



# Python Booleans

**布林型態**：

僅有True和False兩個值。

**特別用法**:
- 以數值表示booleans，0為「偽」、非0為「真」
- 以字串表示booleans，空字串為「偽」、非空字串為「真」

**Notice**: ' '為「一個空白組成的字串」，而非「空字串」('')

In [15]:
print(bool(0))
print(bool(1))
print(bool(46235))
print(bool(-1))
print(bool(-3254236))

False
True
True
True
True


In [16]:
print(bool(''))
print(bool('shwh'))

False
True


In [17]:
print(bool(' '))

True


# Python Operators

**運算子**:

用於表示運算的符號。

常見運算符號有：
- \+ 加
- \- 減
- \* 乘
- \/ 除
- \% 餘數
- \** or \^ 次方
- \// 商數
- \& 且
- \| 或

位元運算:
- \>> 位元向右平移
- \<< 位元向左平移

In [18]:
print(3+2)
print(3-2)
print(3*2)
print(3/2)
print(3%2)
print(3**2)
print(3//2)
print(True & False)
print(True | False)

5
1
6
1.5
1
9
1
False
True


In [19]:
print("嗨"+"囉"+"！")
print("嗨"*3)

嗨囉！
嗨嗨嗨


In [20]:
print("嗨" & "囉")

TypeError: unsupported operand type(s) for &: 'str' and 'str'

In [21]:
print("嗨嗨" - "嗨")

TypeError: unsupported operand type(s) for -: 'str' and 'str'

## 位元平移運算子

在電腦中，所有數值皆是以二進制 (0, 1) 表式，例如：

$(2)_{10} = (00000010)_{2}$

$(5)_{10} = (00000101)_{2}$

「2>>1」表示，將2的位元值向右平移1單位。即$(00000010)_{2}$變為$(0000001)_{2}$。此時，因為平移後，最左邊少了「1」位元，直接補0來表示一個完整值。所以，「2>>1」運算結果為$(00000001)_{2}$，此值又等於$(1)_{10}$。

由上述可知，
- Ex1:「5>>1」等於$(00000101)_{2}$ >> 1，則為$(00000010)_{2}$，結果等於$(2)_{10}$。
- Ex2:「5>>2」等於$(00000101)_{2}$ >> 2，則為$(00000001)_{2}$，結果等於$(1)_{10}$。

**補充**:
可以使用bin()函數將值轉為二進制。

In [22]:
print(bin(2))
print(2>>1)

0b10
1


In [23]:
print(bin(5))
print(5>>1)

0b101
2


In [24]:
print(5>>2)

1


**補充**:
由上述例子可知，在正整數下「x>>1」與「x//$2^{1}$」可獲得相同結果、「x>>2」與「x//$2^{2}$」可獲得相同結果。

可透過相同方法得到，「整數」透過位元運算子，皆可獲得相同結果。

在IEEE-754浮點數表示法中，定義了正數、負數、小數與指數表示法。而位元運算子受到此表示法的定義造成運算上的差異，若想鑽研可讀相關文章。

[Reference]()

- [從 IEEE 754 標準來看為什麼浮點誤差是無法避免的](https://medium.com/starbugs/see-why-floating-point-error-can-not-be-avoided-from-ieee-754-809720b32175)

# Python Casting

**轉型**:

將資料在不同資料型態中轉換，例如：整數轉為字串、浮點數轉為整數。

In [25]:
x = 1.3
print(type(x))
print(int(x))
print(type(x))
x = str(x)
print(x)
print(type(x))

<class 'float'>
1
<class 'float'>
1.3
<class 'str'>
